daffodil-io/src/main/scala/org/apache/daffodil/io/DataOutputStreamImplMixin.scala

```diff
@@ -657,6 +668,64 @@ trait DataOutputStreamImplMixin extends DataStreamCommonState
     }
   }
 
+  final def putFile(
+    path: Path,
+    lengthInBits: Long,
+    chunkSizeInBytes: Int,
+    finfo: FormatInfo): Long = {
+
+    val fileStream =
+      try {
+        Files.newInputStream(path, StandardOpenOption.READ)
+      } catch {
+        case e: Exception =>
+          throw new BlobIOException(
+            "Unable to open BLOB %s for reading: %s".format(
+              path.toString,
+              e.getMessage()))
+      }
+
+    val array = new Array[Byte](chunkSizeInBytes)
+
+    val chunkSizeInBits = chunkSizeInBytes * 8
+
+    var remainingBitsToPut = lengthInBits
+    var fileHasData = true
+    var bitsWritten: Long = 0
+    while (remainingBitsToPut > 0 && fileHasData) {
+      val bitsToRead = Math.min(remainingBitsToPut, chunkSizeInBits)
+      val bytesToRead = (bitsToRead + 7) / 8
+      val bytesRead = fileStream.read(array, 0, bytesToRead.toInt)
+      if (bytesRead == -1) {
+        fileHasData = false
+      } else {
+        val bitsToPut = Math.min(bytesRead * 8, bitsToRead)
+        val ret = putByteArray(array, bitsToPut.toInt, finfo, ignoreByteOrder=true)
+        if (!ret) {
+          fileStream.close()
+          throw new BlobIOException(
+            "Failed to write BLOB data: %s".format(path.toString))
```

Comment:
Is this really about BLOB? I thought this would apply any time the buffer grows too big. You probably want an I/O Exception certainly, but not one specific to BLOB.

Reply:
Meant to change this, but completely forgot before I pushed this up for review.  Will fix.